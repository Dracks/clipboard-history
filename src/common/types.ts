import { Config } from "./config";

export type WindowPage = "config" | "Debug"

export type PageData = {
    "config": Config,
    "Debug": String
}

export type PageDataSend = {
    page: WindowPage,
    data: Config | String
}

export enum EventsName {
    Log = "log",
    Error = "error",
    Load = "load",
    Save = "save",
}

export enum ChangeContext{
    // When there is a new clipboard detected
    new = "new",
    // If the change is provoqued by a shortcut
    shortcut = "shortcut",
    // If it's selected from the tray
    manual = "manual",
    // If is when the application starts
    start = "start"
}