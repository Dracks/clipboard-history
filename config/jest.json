{
    "moduleFileExtensions": [
        "js",
        "json",
        "ts",
        "tsx"
    ],
    "rootDir": "../src",
    "testRegex": ".spec.tsx?$",
    "transform": {
        "^.+\\.tsx?$": "ts-jest"
    },
    "coverageDirectory": "../coverage",
    "testEnvironment": "jsdom",
    "globals": {
        "ts-jest": {
          "tsConfig": "./config/tsconfig.json"
        }
    }
}