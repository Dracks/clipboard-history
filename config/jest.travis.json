{
    "moduleFileExtensions": [
        "js",
        "json",
        "ts"
    ],
    "rootDir": "../src",
    "testRegex": ".spec.ts$",
    "transform": {
        "^.+\\.ts$": "ts-jest"
    },
    "testEnvironment": "node",
    "collectCoverage": true,
    "coverageReporters":["text-lcov"]
}