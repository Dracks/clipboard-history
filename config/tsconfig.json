{
    "compilerOptions": {
        "rootDir": "../src/",
        "module": "commonjs",
        "moduleResolution": "node",
        "target": "es6",
        "strict": true,
        "esModuleInterop": true,
        "sourceMap": true,
        "baseUrl": ".",
        "jsx": "react",
        "paths": {
          "*": ["node_modules/*"]
        },
        "types": [
          "node",
          "electron",
          "jest"
        ]
    },
}