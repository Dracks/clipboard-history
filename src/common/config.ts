import { ChangeContext } from "./types";

export type NotificationSystem = 'electron' | 'node';

export type KeyboardShortcut = string

export interface ShortcutsConfig {
    next: KeyboardShortcut,
    previous: KeyboardShortcut,
    removeCurrent: KeyboardShortcut
}

export interface Config {
    historyLength: number,
    notifications: {
        type?: NotificationSystem,
    } & {
        [key in ChangeContext]: boolean
    },
    shortcuts: ShortcutsConfig
}
