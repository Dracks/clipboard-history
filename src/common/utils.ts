
export const mergeObjects = (obj1: any, obj2: any)=>{
    if (obj2 instanceof Array){
        return obj2
    }
    const newProps = Object.keys(obj2)
        .reduce((ac, key)=>{
            let obj2Prop = obj2[key]
            if (typeof obj2Prop === "object" ){
                obj2Prop = mergeObjects(obj1[key] || {}, obj2Prop)
            }
            return {
                ...ac,
                [key]: obj2Prop,
            }
        }, {} as any)
    return {
        ...obj1,
        ...newProps,
    }
}

export type ConstructorClass<O extends {}, T extends {}> = {new(options: O):T}
