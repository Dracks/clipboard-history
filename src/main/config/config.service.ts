import { Config, NotificationSystem } from "../../common/config";
import { ChangeContext } from "../../common/types";
import DataBase from "../core/db";
import { ClipboardEventEmitter, ClipboardEventEnum } from "../types";
import WindowManager from '../window/window.manager';

export const initialConfig : Config = {
    historyLength: 15,
    notifications: {
        type: 'electron',
        new: false,
        shortcut: true,
        manual: false,
        start: false
    }
}

class ConfigService{
    private config: Config

    get maxHistory(): number {
        return this.config.historyLength
    }

    get typeNotifications(): NotificationSystem | undefined{
        return this.config.notifications.type
    }

    get selectedContextNotifications(): Array<ChangeContext>{
        const configNotifications = this.config.notifications
        return Object
            .values(ChangeContext)
            .filter((key:ChangeContext)=>configNotifications[key])
    }

    constructor(private bus: ClipboardEventEmitter, private db: DataBase<Config>, private wm: WindowManager){
        this.config = this.db.read()
        this.bus.on(ClipboardEventEnum.EditConfig, this.openEdit.bind(this))
    }

    openEdit(){
        const data = this.wm.createSingleInstance("config", this.config)
        if (data){
            data.subscribe((d)=>{
                this.config = d;
                this.save()
            })
        }
    }

    save(){
        this.db.write(this.config)
        this.bus.emit(ClipboardEventEnum.ConfigChanged)
    }
}

export default ConfigService