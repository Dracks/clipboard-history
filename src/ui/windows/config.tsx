import React, { useCallback } from 'react';
import { Config, ShortcutsConfig } from "../../common/config";
import { ChangeContext } from '../../common/types';
import { mergeObjects } from '../../common/utils';
import { Checkbox, Page, Section, Spacer, TextField, useModal } from '../theme';
import { getGeneric } from '../utils/utils';
import { WindowPageProps } from './interface';
import SetShortcut from './set_shortcut';


const NOTIFICATIONS_EXPLANATIONS : {
    [key in ChangeContext]: string
} = {
    manual: "When some clipboard is selected from the tray",
    new: "When the clipboard changed",
    shortcut: "When the change is provoqued by a shortcut",
    start: "When the application starts"
}

const SHORTCUTS_EXPLANATIONS : {
    [key in keyof ShortcutsConfig]: string
} = {
    next: "The shortcut to select next value",
    previous: "The shortcut to select previous value",
    removeCurrent: "The shortcut to remove the current value from the history and move to next"
}

const Configuration = ({data, save}: WindowPageProps<Config>)=>{
    const config = data
    const setConfig = save
    const update = useCallback((newData: Partial<Config>)=>{
        setConfig(mergeObjects(config, newData))
    }, [config])
    const shortcuts = Object.keys(config.shortcuts) as any

    return <Page >
            <Section title="General" >
                <h2>History length</h2>
                <TextField.Number {...getGeneric(["historyLength"], config, update)}/>
                <Spacer />
                {shortcuts.map((key: keyof ShortcutsConfig)=>{
                    const {value, onChange} = getGeneric(["shortcuts", key], config, ()=>undefined)
                    const {modal, show} = useModal(({close})=><SetShortcut save={(shortcut)=>{onChange(shortcut);close()}}/>, key)
                    return <div key={key}>
                        {modal}
                        <div>{SHORTCUTS_EXPLANATIONS[key]}</div>
                        <TextField.String
                            value={value}
                            onClick={show}
                            disabled
                            />
                        <Spacer />
                    </div>
                })}
            </Section>
            <Spacer/>
            <Section title="notifications">
                {Object.values(ChangeContext).map((key:ChangeContext)=>(
                    <div key={key}>
                        <Checkbox
                            id={key}
                            text={NOTIFICATIONS_EXPLANATIONS[key]}
                            {...getGeneric(["notifications", key], config, update)}
                            />
                    </div>
                ))}
            </Section>
        </Page>
}

export default Configuration