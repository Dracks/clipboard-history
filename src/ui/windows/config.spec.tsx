import { fireEvent, render, RenderResult } from '@testing-library/react'
import React from 'react'
import { Config, initialConfig } from '../../common/config'
import Subject from './config'


describe('ui/windows/config', ()=>{
    let onSave: jest.Mock
    let query: RenderResult

    beforeEach(()=>{
        onSave = jest.fn()

        //euiIconSpied = jest.spyOn(EuiIcon as any, "loadIconComponent").mockReturnValue(null)
    })
    const mount = (data: Config)=>{
        query = render(<Subject data={data} save={onSave}/>)
    }

    afterEach(()=>{
        query.unmount()
    })

    it('Renders well', ()=>{
        mount(initialConfig)
        expect(query.container).toMatchSnapshot()
    })

    it('Restore defaults', ()=>{
        const config : Config = {
            ...initialConfig,
            shortcuts: {
                ...initialConfig.shortcuts,
                next: "Ping Pollo"
            }
        }
        mount(config)
        expect(query.container).toMatchSnapshot("Button refresh not disabled")

        const button = query.getAllByRole('button').filter(e=>e.getAttribute('disabled')===null)
        expect(button).toHaveLength(1)
        fireEvent.click(button[0])
        expect(onSave).toBeCalledTimes(1)
        expect(onSave).toBeCalledWith(initialConfig)
    })

    it('Change configuration for the tray', ()=>{
        mount(initialConfig)

        // Tray length
        let search = query.getByText('Length show in tray')
        const input = search.nextElementSibling!.firstChild!.firstChild as HTMLInputElement
        fireEvent.change(input, {target: {value: "30"}})

        expect(onSave).toBeCalledTimes(1)
        expect(onSave).toBeCalledWith({
            ...initialConfig,
            tray: {
                trim: false,
                size: 30
            }
        })

        // Trim tray
        search = query.getByText('Trim clipboard values')
        const checkbox = search.previousElementSibling!.previousElementSibling as HTMLInputElement
        fireEvent.click(checkbox)

        expect(onSave).toBeCalledTimes(2)
        expect(onSave).toBeCalledWith({
            ...initialConfig,
            tray: {
                trim: true,
                size: 20
            }
        })
    })
})
