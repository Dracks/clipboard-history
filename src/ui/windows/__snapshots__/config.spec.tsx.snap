// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/windows/config Renders well 1`] = `
<div>
  <div
    class="euiPage"
  >
    <div
      class="euiPageBody"
    >
      <div
        class="euiPanel euiPanel--paddingMedium euiPanel--hasBetaBadge"
      >
        <span
          class="euiPanel__betaBadgeWrapper"
        >
          <span
            class="euiBetaBadge euiPanel__betaBadge"
            title="General"
          >
            General
          </span>
        </span>
        <h2>
          History length
        </h2>
        <div
          class="euiFormControlLayout"
        >
          <div
            class="euiFormControlLayout__childrenWrapper"
          >
            <input
              class="euiFieldNumber"
              type="number"
              value="15"
            />
          </div>
        </div>
        <div
          class="euiSpacer euiSpacer--l"
        />
        <div>
          <div>
            The shortcut to select next value
          </div>
          <div
            style="display: flex;"
          >
            <div
              style="flex-grow: 1;"
            >
              <div>
                <div
                  class="euiFormControlLayout euiFormControlLayout--readOnly"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <input
                      class="euiFieldText"
                      readonly=""
                      type="text"
                      value="CommandOrControl+F12"
                    />
                  </div>
                </div>
              </div>
            </div>
            <button
              aria-label="refresh"
              class="euiButtonIcon euiButtonIcon--primary"
              disabled=""
              type="button"
            >
              <div
                title="icon"
              >
                {"className":"euiButtonIcon__icon","type":"refresh","size":"m","aria-hidden":"true"}
              </div>
            </button>
          </div>
        </div>
        <div>
          <div>
            The shortcut to select previous value
          </div>
          <div
            style="display: flex;"
          >
            <div
              style="flex-grow: 1;"
            >
              <div>
                <div
                  class="euiFormControlLayout euiFormControlLayout--readOnly"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <input
                      class="euiFieldText"
                      readonly=""
                      type="text"
                      value="CommandOrControl+F11"
                    />
                  </div>
                </div>
              </div>
            </div>
            <button
              aria-label="refresh"
              class="euiButtonIcon euiButtonIcon--primary"
              disabled=""
              type="button"
            >
              <div
                title="icon"
              >
                {"className":"euiButtonIcon__icon","type":"refresh","size":"m","aria-hidden":"true"}
              </div>
            </button>
          </div>
        </div>
        <div>
          <div>
            The shortcut to remove the current value from the history and move to next
          </div>
          <div
            style="display: flex;"
          >
            <div
              style="flex-grow: 1;"
            >
              <div>
                <div
                  class="euiFormControlLayout euiFormControlLayout--readOnly"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <input
                      class="euiFieldText"
                      readonly=""
                      type="text"
                      value="CommandOrControl+F9"
                    />
                  </div>
                </div>
              </div>
            </div>
            <button
              aria-label="refresh"
              class="euiButtonIcon euiButtonIcon--primary"
              disabled=""
              type="button"
            >
              <div
                title="icon"
              >
                {"className":"euiButtonIcon__icon","type":"refresh","size":"m","aria-hidden":"true"}
              </div>
            </button>
          </div>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiPanel euiPanel--paddingMedium euiPanel--hasBetaBadge"
      >
        <span
          class="euiPanel__betaBadgeWrapper"
        >
          <span
            class="euiBetaBadge euiPanel__betaBadge"
            title="Tray"
          >
            Tray
          </span>
        </span>
        <h2>
          Length show in tray
        </h2>
        <div
          class="euiFormControlLayout"
        >
          <div
            class="euiFormControlLayout__childrenWrapper"
          >
            <input
              class="euiFieldNumber"
              type="number"
              value="20"
            />
          </div>
        </div>
        <div
          class="euiSpacer euiSpacer--l"
        />
        <div
          class="euiCheckbox"
        >
          <input
            class="euiCheckbox__input"
            id="tray.trim"
            type="checkbox"
          />
          <div
            class="euiCheckbox__square"
          />
          <label
            class="euiCheckbox__label"
            for="tray.trim"
          >
            Trim clipboard values
          </label>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiPanel euiPanel--paddingMedium euiPanel--hasBetaBadge"
      >
        <span
          class="euiPanel__betaBadgeWrapper"
        >
          <span
            class="euiBetaBadge euiPanel__betaBadge"
            title="notifications"
          >
            notifications
          </span>
        </span>
        <div>
          <div
            class="euiCheckbox"
          >
            <input
              class="euiCheckbox__input"
              id="new"
              type="checkbox"
            />
            <div
              class="euiCheckbox__square"
            />
            <label
              class="euiCheckbox__label"
              for="new"
            >
              When the clipboard changed
            </label>
          </div>
        </div>
        <div>
          <div
            class="euiCheckbox"
          >
            <input
              checked=""
              class="euiCheckbox__input"
              id="shortcut"
              type="checkbox"
            />
            <div
              class="euiCheckbox__square"
            />
            <label
              class="euiCheckbox__label"
              for="shortcut"
            >
              When the change is provoqued by a shortcut
            </label>
          </div>
        </div>
        <div>
          <div
            class="euiCheckbox"
          >
            <input
              class="euiCheckbox__input"
              id="manual"
              type="checkbox"
            />
            <div
              class="euiCheckbox__square"
            />
            <label
              class="euiCheckbox__label"
              for="manual"
            >
              When some clipboard is selected from the tray
            </label>
          </div>
        </div>
        <div>
          <div
            class="euiCheckbox"
          >
            <input
              class="euiCheckbox__input"
              id="start"
              type="checkbox"
            />
            <div
              class="euiCheckbox__square"
            />
            <label
              class="euiCheckbox__label"
              for="start"
            >
              When the application starts
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ui/windows/config Restore defaults: Button refresh not disabled 1`] = `
<div>
  <div
    class="euiPage"
  >
    <div
      class="euiPageBody"
    >
      <div
        class="euiPanel euiPanel--paddingMedium euiPanel--hasBetaBadge"
      >
        <span
          class="euiPanel__betaBadgeWrapper"
        >
          <span
            class="euiBetaBadge euiPanel__betaBadge"
            title="General"
          >
            General
          </span>
        </span>
        <h2>
          History length
        </h2>
        <div
          class="euiFormControlLayout"
        >
          <div
            class="euiFormControlLayout__childrenWrapper"
          >
            <input
              class="euiFieldNumber"
              type="number"
              value="15"
            />
          </div>
        </div>
        <div
          class="euiSpacer euiSpacer--l"
        />
        <div>
          <div>
            The shortcut to select next value
          </div>
          <div
            style="display: flex;"
          >
            <div
              style="flex-grow: 1;"
            >
              <div>
                <div
                  class="euiFormControlLayout euiFormControlLayout--readOnly"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <input
                      class="euiFieldText"
                      readonly=""
                      type="text"
                      value="Ping Pollo"
                    />
                  </div>
                </div>
              </div>
            </div>
            <button
              aria-label="refresh"
              class="euiButtonIcon euiButtonIcon--primary"
              type="button"
            >
              <div
                title="icon"
              >
                {"className":"euiButtonIcon__icon","type":"refresh","size":"m","aria-hidden":"true"}
              </div>
            </button>
          </div>
        </div>
        <div>
          <div>
            The shortcut to select previous value
          </div>
          <div
            style="display: flex;"
          >
            <div
              style="flex-grow: 1;"
            >
              <div>
                <div
                  class="euiFormControlLayout euiFormControlLayout--readOnly"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <input
                      class="euiFieldText"
                      readonly=""
                      type="text"
                      value="CommandOrControl+F11"
                    />
                  </div>
                </div>
              </div>
            </div>
            <button
              aria-label="refresh"
              class="euiButtonIcon euiButtonIcon--primary"
              disabled=""
              type="button"
            >
              <div
                title="icon"
              >
                {"className":"euiButtonIcon__icon","type":"refresh","size":"m","aria-hidden":"true"}
              </div>
            </button>
          </div>
        </div>
        <div>
          <div>
            The shortcut to remove the current value from the history and move to next
          </div>
          <div
            style="display: flex;"
          >
            <div
              style="flex-grow: 1;"
            >
              <div>
                <div
                  class="euiFormControlLayout euiFormControlLayout--readOnly"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <input
                      class="euiFieldText"
                      readonly=""
                      type="text"
                      value="CommandOrControl+F9"
                    />
                  </div>
                </div>
              </div>
            </div>
            <button
              aria-label="refresh"
              class="euiButtonIcon euiButtonIcon--primary"
              disabled=""
              type="button"
            >
              <div
                title="icon"
              >
                {"className":"euiButtonIcon__icon","type":"refresh","size":"m","aria-hidden":"true"}
              </div>
            </button>
          </div>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiPanel euiPanel--paddingMedium euiPanel--hasBetaBadge"
      >
        <span
          class="euiPanel__betaBadgeWrapper"
        >
          <span
            class="euiBetaBadge euiPanel__betaBadge"
            title="Tray"
          >
            Tray
          </span>
        </span>
        <h2>
          Length show in tray
        </h2>
        <div
          class="euiFormControlLayout"
        >
          <div
            class="euiFormControlLayout__childrenWrapper"
          >
            <input
              class="euiFieldNumber"
              type="number"
              value="20"
            />
          </div>
        </div>
        <div
          class="euiSpacer euiSpacer--l"
        />
        <div
          class="euiCheckbox"
        >
          <input
            class="euiCheckbox__input"
            id="tray.trim"
            type="checkbox"
          />
          <div
            class="euiCheckbox__square"
          />
          <label
            class="euiCheckbox__label"
            for="tray.trim"
          >
            Trim clipboard values
          </label>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiPanel euiPanel--paddingMedium euiPanel--hasBetaBadge"
      >
        <span
          class="euiPanel__betaBadgeWrapper"
        >
          <span
            class="euiBetaBadge euiPanel__betaBadge"
            title="notifications"
          >
            notifications
          </span>
        </span>
        <div>
          <div
            class="euiCheckbox"
          >
            <input
              class="euiCheckbox__input"
              id="new"
              type="checkbox"
            />
            <div
              class="euiCheckbox__square"
            />
            <label
              class="euiCheckbox__label"
              for="new"
            >
              When the clipboard changed
            </label>
          </div>
        </div>
        <div>
          <div
            class="euiCheckbox"
          >
            <input
              checked=""
              class="euiCheckbox__input"
              id="shortcut"
              type="checkbox"
            />
            <div
              class="euiCheckbox__square"
            />
            <label
              class="euiCheckbox__label"
              for="shortcut"
            >
              When the change is provoqued by a shortcut
            </label>
          </div>
        </div>
        <div>
          <div
            class="euiCheckbox"
          >
            <input
              class="euiCheckbox__input"
              id="manual"
              type="checkbox"
            />
            <div
              class="euiCheckbox__square"
            />
            <label
              class="euiCheckbox__label"
              for="manual"
            >
              When some clipboard is selected from the tray
            </label>
          </div>
        </div>
        <div>
          <div
            class="euiCheckbox"
          >
            <input
              class="euiCheckbox__input"
              id="start"
              type="checkbox"
            />
            <div
              class="euiCheckbox__square"
            />
            <label
              class="euiCheckbox__label"
              for="start"
            >
              When the application starts
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
